File: backend/API/serializers/question.py
from rest_framework import serializers
from neomodel.exceptions import DoesNotExist
from ..serializers import BaseSerializer
from ..models import Question, Theme
from ..errors import NotFound


class QuestionSerializer(BaseSerializer):
    """
    Sérializer du node Question
    """

    texte = serializers.CharField(required=True)

    # Inputs
    input_fields = {"theme_uuid": {"relationship": "theme", "node": Theme}}
    theme_uuid = serializers.CharField(write_only=True, required=False)

    # Outputs
    theme = serializers.SerializerMethodField(read_only=True)
    extraits = serializers.SerializerMethodField(read_only=True)

    def __init__(self, *args, **kwargs):
        super().__init__(Question, *args, **kwargs)

    def get_theme(self, question):
        """
        Renvoie un lien propre vers le theme :
        """
        theme = question.theme.single()
        return (
            self.get_url("theme-detail", kwargs={"uuid": theme.uuid}) if theme else None
        )

    def get_extraits(self, question):
        """
        Renvoie un lien propre vers les extraits :
        """
        return self.get_url("extrait-list", kwargs={"question_uuid": question.uuid})

File: frontend/src/model/question.js
import Model from "@model/model.js";
import Theme from "@model/theme.js";
import Extrait from "@model/extrait.js";

export default class Question extends Model {
    #texte;
    #theme;
    #extraits;
    #theme_uuid;

    constructor({ uuid, texte, theme, extraits }) {
        super(uuid);
        this.#texte = texte;
        this.#theme = theme;
        this.#extraits = extraits;
        this.#theme_uuid = null;
    }

    static get endpoint() { return "questions"; }

    get texte() { return this.#texte; }
    set texte(value) { this.#texte = Question.validateString(value, "texte"); }

    get theme() { return this.fetchDetail(this.#theme, Theme); }
    set theme(value) { this.#theme_uuid = Question.validateString(value, "theme_uuid"); }

    async extraits(args) { return await this.fetchList(this.#extraits, Extrait, args); }

    fromJSON(json) {
        super.fromJSON(json);
        this.#texte = json.texte;
        this.#theme = json.theme;
        this.#extraits = json.extraits;
        return this;
    }

    toJSON() {
        return {
            uuid: this.uuid,
            texte: this.#texte,
            theme_uuid: this.#theme_uuid
        };
    }
}
